docs/source/optim.rst

```diff
@@ -303,6 +314,76 @@ algorithms.
     lr_scheduler.OneCycleLR
     lr_scheduler.CosineAnnealingWarmRestarts
 
+How to utilize named parameters to load optimizer state dict
+--------------------------------------------
+
+The function :func:`~Optimizer.load_state_dict` ignores the optional ``param_names`` input in the
+loaded state dict if present (to maintain compatibility).
+To utilize the loaded parameters names from the loaded state dict, a custom ``register_load_state_dict_pre_hook``
+needs to be implemented.
+
+This can be useful, for instance, when the model architecture changes, but the weights and states need to
+remain unchanged. The following example demonstrates how to implement this customization.
+
+Example::
+
+    class OneLayerModel(nn.Module):
+        def __init__(self):
+            super(OneLayerModel, self).__init__()
+            self.fc = nn.Linear(3, 4)
+
+        def forward(self, x):
+            return self.fc(x)
+
+    model = OneLayerModel()
+    optimizer = optim.SGD(model.named_parameters(), lr=0.01, momentum=0.9)
+    torch.save(optimizer.state_dict(), PATH)
+
+Assuming that for the following ``model2`` we wish to load the states of ``fc`` of ``model`` into both ``fc1`` and ``fc2`` of ``model2``::
```

Comment:
Do you have a concrete example of when someone wants to do this?

Higher level question: I would imagine a better example for registering a custom hook is to maintain the same names as the previous state_dict. Is that already the default behavior with no hook --> is there a test below that tests this?

Or you can answer this q: what happens without the custom hook? Will the param_names get copied over or will they get completely ignored (the new state_dict no longer has them)?